fix(postgres): fix array_index operation

krzysztof-kwitt opened the pull request:
Hi.

Small fix for Postgres backend.
Now the `ibis/backends/tests/test_array.py` tests for Postgres look like below
```
ibis/backends/tests/test_array.py::test_array_discovery_clickhouse[postgres] XFAIL (backend supports nullable nested types)           [  5%]
ibis/backends/tests/test_array.py::test_array_scalar[postgres] PASSED                                                                 [ 11%]
ibis/backends/tests/test_array.py::test_list_literal[postgres] PASSED                                                                 [ 16%]
ibis/backends/tests/test_array.py::test_array_concat[postgres] PASSED                                                                 [ 22%]
ibis/backends/tests/test_array.py::test_unnest_no_nulls[postgres] PASSED                                                              [ 27%]
ibis/backends/tests/test_array.py::test_unnest_default_name[postgres] PASSED                                                          [ 33%]
ibis/backends/tests/test_array.py::test_unnest_simple[postgres] PASSED                                                                [ 38%]
ibis/backends/tests/test_array.py::test_array_index[postgres-2] PASSED                                                                [ 44%]
ibis/backends/tests/test_array.py::test_array_index[postgres-1] PASSED                                                                [ 50%]
ibis/backends/tests/test_array.py::test_unnest_idempotent[postgres] PASSED                                                            [ 55%]
ibis/backends/tests/test_array.py::test_array_index[postgres-0] PASSED                                                                [ 61%]
ibis/backends/tests/test_array.py::test_array_discovery_desired[postgres] XFAIL (backend does not support nullable nested types)      [ 66%]
ibis/backends/tests/test_array.py::test_array_column[postgres] PASSED                                                                 [ 72%]
ibis/backends/tests/test_array.py::test_np_array_literal[postgres] PASSED                                                             [ 77%]
ibis/backends/tests/test_array.py::test_array_discovery_snowflake[postgres] XFAIL (backend does not implement arrays like snowflake)  [ 83%]
ibis/backends/tests/test_array.py::test_unnest_complex[postgres] PASSED                                                               [ 88%]
ibis/backends/tests/test_array.py::test_array_length[postgres] PASSED                                                                 [ 94%]
ibis/backends/tests/test_array.py::test_array_discovery_postgres_duckdb[postgres] PASSED                                              [100%]
```
Best regards,
Kamil

cpcloud commented:
Awesome! Can you change the commit prefix to `fix` instead of `bugfix`? Then it should be good to merge!

cpcloud commented:
Oh also, the character after the colon should be lowercase!

krzysztof-kwitt commented:
@cpcloud I fixed it. I also rewrote it a bit because the problem only occurs when the array is created and then we lookup index.
Incorrect SQL:
```
SELECT [1,2,3][0]
```
Correct SQL:
```
SELECT ([1,2,3])[0]
```

cpcloud commented on ibis/backends/postgres/registry.py:
```diff
@@ -431,6 +431,14 @@ def _array_slice(t, op):
+def _array_index(t, op):
+    sa_array = t.translate(op.arg)
+    sa_index = t.translate(op.index)
+    if isinstance(op.arg, ops.Literal) and op.arg.dtype.is_array():
```
No need to check `is_array` here, since this is true by definition of `op.arg`.

krzysztof-kwitt replied:
You're right. Fixed.

cpcloud reviewed (approved):
Awesome!
